### PR TITLE
KAFKA-15598 & KAFKA-15461: Add integration tests for DescribeGroups API, DeleteGroups API, OffsetDelete API and ListGroups API

### PR DESCRIPTION
**Pull request by dongnuo123**
This pr includes integration tests for four group coordinator APIs with new group coordinator and new protocol, new group coordinator and old protocol, and old group coordinator and old protocols.
- DescribeGroups
- DeleteGroups
- OffsetDelete
- ListGroups

### Jira
https://issues.apache.org/jira/browse/KAFKA-15598
https://issues.apache.org/jira/browse/KAFKA-15461

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/test/scala/unit/kafka/server/OffsetDeleteRequestTest.scala [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package kafka.server
+
+
```
Comment (rreddy-22) [inline core/src/test/scala/unit/kafka/server/OffsetDeleteRequestTest.scala]: nit: extra line

File: core/src/test/scala/unit/kafka/server/DeleteGroupsRequestTest.scala [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+          .setGroupState(if (useNewProtocol) ConsumerGroupState.DEAD.toString else GenericGroupState.DEAD.toString)
+        ),
+        describeGroups(
+          groupIds = List("grp"),
+          version = version.toShort
+        )
+      )
+    }
+  }
+}
```
Comment (rreddy-22) [inline core/src/test/scala/unit/kafka/server/DeleteGroupsRequestTest.scala:135]: nit: add line at the end of the file

File: core/src/test/scala/unit/kafka/server/DeleteGroupsRequestTest.scala [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    new ClusterConfigProperty(key = "unstable.api.versions.enable", value = "false"),
+    new ClusterConfigProperty(key = "group.coordinator.new.enable", value = "false"),
+    new ClusterConfigProperty(key = "offsets.topic.num.partitions", value = "1"),
+    new ClusterConfigProperty(key = "offsets.topic.replication.factor", value = "1")
+  ))
+  def testDeleteGroupsWithOldConsumerGroupProtocolAndOldGroupCoordinator(): Unit = {
+    testDeleteGroups(false)
+  }
+
+  private def testDeleteGroups(useNewProtocol: Boolean): Unit = {
```
Comment (rreddy-22) [inline core/src/test/scala/unit/kafka/server/DeleteGroupsRequestTest.scala:66]: Can we add a comment here or where we create the empty groups for deletion that we're testing for both a non-empty group and an empty group? Like "Testing delete group on empty and non-empty groups"

File: core/src/test/scala/unit/kafka/server/DescribeGroupsRequestTest.scala [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    testDescribeGroups()
+  }
+
+  @ClusterTest(clusterType = Type.ALL, serverProperties = Array(
+    new ClusterConfigProperty(key = "unstable.api.versions.enable", value = "false"),
+    new ClusterConfigProperty(key = "group.coordinator.new.enable", value = "false"),
+    new ClusterConfigProperty(key = "offsets.topic.num.partitions", value = "1"),
+    new ClusterConfigProperty(key = "offsets.topic.replication.factor", value = "1")
+  ))
+  def testDescribeGroupsWithOldConsumerGroupProtocolAndOldGroupCoordinator(): Unit = {
```
Comment (rreddy-22) [inline core/src/test/scala/unit/kafka/server/DescribeGroupsRequestTest.scala:50]: Are we going to test new consumer group protocol and new group coordinator for this api as well?

File: core/src/test/scala/unit/kafka/server/DescribeGroupsRequestTest.scala [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+            .setGroupState(GenericGroupState.DEAD.toString)
+        ),
+        describeGroups(
+          groupIds = List("grp-1", "grp-2","grp-unknown"),
+          version = version.toShort
+        )
+      )
+    }
+  }
+}
```
Comment (rreddy-22) [inline core/src/test/scala/unit/kafka/server/DescribeGroupsRequestTest.scala:119]: nit:line

File: core/src/test/scala/unit/kafka/server/GroupCoordinatorBaseRequestTest.scala [hunk 6]
```diff
@@ -181,6 +185,52 @@ class GroupCoordinatorBaseRequestTest(cluster: ClusterInstance) {
...
+  protected def deleteOffset(
+    groupId: String,
+    topic: String,
+    partition: Int,
+    expectedResponseError: Errors = Errors.NONE,
+    expectedPartitionError: Errors = Errors.NONE,
+    version: Short
+  ): Unit = {
+    if (expectedResponseError != Errors.NONE && expectedPartitionError != Errors.NONE) {
+      fail("deleteOffset: neither expectedResponseError nor expectedTopicError is none.")
```
Comment (rreddy-22) [inline core/src/test/scala/unit/kafka/server/GroupCoordinatorBaseRequestTest.scala]: nit: NONE* is the message

File: core/src/test/scala/unit/kafka/server/OffsetDeleteRequestTest.scala [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+      deleteOffset(
+        groupId = "",
+        topic = "foo",
+        partition = 2,
+        expectedResponseError = Errors.INVALID_GROUP_ID,
+        version = version.toShort
+      )
+    }
+  }
+}
```
Comment (rreddy-22) [inline core/src/test/scala/unit/kafka/server/OffsetDeleteRequestTest.scala:182]: nit: line

File: core/src/test/scala/unit/kafka/server/DescribeGroupsRequestTest.scala (hunk 4 above)
Comment (dongnuo123) [reply core/src/test/scala/unit/kafka/server/DescribeGroupsRequestTest.scala:50]: No, because DescribeGroups only works for old protocol. For new protocol we have ConsumerGroupDescribe.

File: core/src/test/scala/unit/kafka/server/DeleteGroupsRequestTest.scala [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    // Join the consumer group. Note that we don't heartbeat here so we must use
+    // a session long enough for the duration of the test.
+    // We test DeleteGroups on empty and non-empty groups. Here we create the non-empty group.
+    joinConsumerGroup(
+      groupId = "grp-non-empty",
+      useNewProtocol = useNewProtocol
+    )
+
+    for (version <- ApiKeys.DELETE_GROUPS.oldestVersion() to ApiKeys.DELETE_GROUPS.latestVersion(isUnstableApiEnabled)) {
+
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/DeleteGroupsRequestTest.scala]: nit: We could remove this empty line.

File: core/src/test/scala/unit/kafka/server/DeleteGroupsRequestTest.scala [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+      // Join the consumer group. Note that we don't heartbeat here so we must use
+      // a session long enough for the duration of the test.
+      val (memberId, memberEpoch) = joinConsumerGroup(
+        groupId = "grp",
+        useNewProtocol = useNewProtocol
+      )
+      leaveGroup(
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/DeleteGroupsRequestTest.scala:98]: nit: Let's add an empty line and a comment explaining why we need to leave the group.

File: core/src/test/scala/unit/kafka/server/DeleteGroupsRequestTest.scala [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+        )
+      } else {
+        assertEquals(
+          List(new DescribedGroup()
+            .setGroupId("grp")
+            .setGroupState(GenericGroupState.DEAD.toString)
+          ),
+          describeGroups(
+            groupIds = List("grp"),
+            version = version.toShort
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/DeleteGroupsRequestTest.scala]: Using `version.toShort` seems incorrect here, isn't it?

File: core/src/test/scala/unit/kafka/server/DescribeGroupsRequestTest.scala [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+import scala.jdk.CollectionConverters._
+
+@Timeout(120)
+@ExtendWith(value = Array(classOf[ClusterTestExtensions]))
+@ClusterTestDefaults(clusterType = Type.KRAFT, brokers = 1)
+@Tag("integration")
+class DescribeGroupsRequestTest(cluster: ClusterInstance) extends GroupCoordinatorBaseRequestTest(cluster) {
+  @ClusterTest(serverProperties = Array(
+    new ClusterConfigProperty(key = "unstable.api.versions.enable", value = "true"),
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/DescribeGroupsRequestTest.scala]: I suppose that we don't need this in this case.

File: core/src/test/scala/unit/kafka/server/DescribeGroupsRequestTest.scala [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+    new ClusterConfigProperty(key = "group.coordinator.new.enable", value = "true"),
+    new ClusterConfigProperty(key = "offsets.topic.num.partitions", value = "1"),
+    new ClusterConfigProperty(key = "offsets.topic.replication.factor", value = "1")
+  ))
+  def testDescribeGroupsWithOldConsumerGroupProtocolAndNewGroupCoordinator(): Unit = {
+    testDescribeGroups()
+  }
+
+  @ClusterTest(clusterType = Type.ALL, serverProperties = Array(
+    new ClusterConfigProperty(key = "unstable.api.versions.enable", value = "false"),
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/DescribeGroupsRequestTest.scala]: ditto.

File: core/src/test/scala/unit/kafka/server/DescribeGroupsRequestTest.scala [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+    new ClusterConfigProperty(key = "group.coordinator.new.enable", value = "false"),
+    new ClusterConfigProperty(key = "offsets.topic.num.partitions", value = "1"),
+    new ClusterConfigProperty(key = "offsets.topic.replication.factor", value = "1")
+  ))
+  def testDescribeGroupsWithOldConsumerGroupProtocolAndOldGroupCoordinator(): Unit = {
+    testDescribeGroups()
+  }
+
+  private def testDescribeGroups(): Unit = {
+
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/DescribeGroupsRequestTest.scala]: nit: We usually don't put an empty line after the method declaration.

File: core/src/test/scala/unit/kafka/server/DescribeGroupsRequestTest.scala [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+    // a session long enough for the duration of the test.
+    val (memberId1, _) = joinConsumerGroupWithOldProtocol(
+      groupId = "grp-1",
+      metadata = Array(1, 2, 3)
+    )
+    val (memberId2, _) = joinConsumerGroupWithOldProtocol(
+      groupId = "grp-2",
+      metadata = Array(1, 2, 3),
+      completeRebalance = false
+    )
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/DescribeGroupsRequestTest.scala]: nit: Let's update the first comment and add one for the second member to explain why we don't complete the rebalance.

File: core/src/test/scala/unit/kafka/server/DescribeGroupsRequestTest.scala [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+            .setGroupState(GenericGroupState.COMPLETING_REBALANCE.toString)
+            .setProtocolType("consumer")
+            .setMembers(List(
+              new DescribedGroupMember()
+                .setMemberId(memberId2)
+                .setGroupInstanceId(null)
+                .setClientId("client-id")
+                .setClientHost("/127.0.0.1")
+            ).asJava),
+          new DescribedGroup()
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/DescribeGroupsRequestTest.scala:108]: nit: Should we add a comment for this case? It is obvious that we should return a Dead group when the group does not exist.

File: core/src/test/scala/unit/kafka/server/DescribeGroupsRequestTest.scala [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+            .setGroupState(GenericGroupState.STABLE.toString)
+            .setProtocolType("consumer")
+            .setProtocolData("consumer-range")
+            .setMembers(List(
+              new DescribedGroupMember()
+                .setMemberId(memberId1)
+                .setGroupInstanceId(null)
+                .setClientId("client-id")
+                .setClientHost("/127.0.0.1")
+                .setMemberMetadata(Array(1, 2, 3))
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/DescribeGroupsRequestTest.scala:92]: Could we also verify memberAssignment?

File: core/src/test/scala/unit/kafka/server/DescribeGroupsRequestTest.scala [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+          new DescribedGroup()
+            .setGroupId("grp-2")
+            .setGroupState(GenericGroupState.COMPLETING_REBALANCE.toString)
+            .setProtocolType("consumer")
+            .setMembers(List(
+              new DescribedGroupMember()
+                .setMemberId(memberId2)
+                .setGroupInstanceId(null)
+                .setClientId("client-id")
+                .setClientHost("/127.0.0.1")
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/DescribeGroupsRequestTest.scala:104]: Should we assert both member metadata and assignment?

File: core/src/test/scala/unit/kafka/server/GroupCoordinatorBaseRequestTest.scala [hunk 18]
```diff
@@ -181,6 +184,52 @@ class GroupCoordinatorBaseRequestTest(cluster: ClusterInstance) {
...
+            .setName(topic)
+            .setPartitions(new OffsetDeleteResponseData.OffsetDeleteResponsePartitionCollection(List(
+              new OffsetDeleteResponseData.OffsetDeleteResponsePartition()
+                .setPartitionIndex(partition)
+                .setErrorCode(expectedPartitionError.code())
+            ).asJava.iterator()))
+        ).asJava.iterator()))
+    } else {
+      expectedResponse
+        .setErrorCode(expectedResponseError.code())
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/GroupCoordinatorBaseRequestTest.scala]: nit: We could bring back `setErrorCode` on the previous line as there is space for it.

File: core/src/test/scala/unit/kafka/server/GroupCoordinatorBaseRequestTest.scala [hunk 19]
```diff
@@ -220,31 +292,92 @@ class GroupCoordinatorBaseRequestTest(cluster: ClusterInstance) {
...
+    version: Short
+  ): List[DescribeGroupsResponseData.DescribedGroup] = {
+    val describeGroupsRequest = new DescribeGroupsRequest.Builder(
+      new DescribeGroupsRequestData()
+        .setGroups(groupIds.asJava)
+    ).build(version)
+
+    val describeGroupsResponse = connectAndReceive[DescribeGroupsResponse](describeGroupsRequest)
+
+    describeGroupsResponse.data().groups().asScala.toList
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/GroupCoordinatorBaseRequestTest.scala]: nit: In Scala, we don't have to put `()` for getters so we usually don't put them in our code base. Here you can use `describeGroupsResponse.data.groups.asScala.toList`.

There are other cases where you could remove them. Could you please look at them?

File: core/src/test/scala/unit/kafka/server/ListGroupsRequestTest.scala [hunk 20]
```diff
@@ -0,0 +1,236 @@
...
+    createOffsetsTopic()
+
+    // Create the topic.
+    createTopic(
+      topic = "foo",
+      numPartitions = 3
+    )
+
+    for (version <- ApiKeys.LIST_GROUPS.oldestVersion() to ApiKeys.LIST_GROUPS.latestVersion(isUnstableApiEnabled)) {
+
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ListGroupsRequestTest.scala:120]: nit: empty line.

File: core/src/test/scala/unit/kafka/server/ListGroupsRequestTest.scala [hunk 21]
```diff
@@ -0,0 +1,236 @@
...
+        listGroups(
+          statesFilter = List.empty,
+          version = version.toShort
+        )
+      )
+    }
+  }
+
+  private def testListGroupsWithOldProtocol(): Unit = {
+
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ListGroupsRequestTest.scala]: nit: Empty line.

File: core/src/test/scala/unit/kafka/server/ListGroupsRequestTest.scala [hunk 22]
```diff
@@ -0,0 +1,236 @@
...
+    createOffsetsTopic()
+
+    // Create the topic.
+    createTopic(
+      topic = "foo",
+      numPartitions = 3
+    )
+
+    for (version <- ApiKeys.LIST_GROUPS.oldestVersion() to ApiKeys.LIST_GROUPS.latestVersion(isUnstableApiEnabled)) {
+
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ListGroupsRequestTest.scala]: ditto.

File: core/src/test/scala/unit/kafka/server/ListGroupsRequestTest.scala [hunk 23]
```diff
@@ -0,0 +1,236 @@
...
+      assertEquals(
+        List.empty,
+        listGroups(
+          statesFilter = List.empty,
+          version = version.toShort
+        )
+      )
+    }
+  }
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ListGroupsRequestTest.scala]: I wonder if we should have both old and new group when we test the new group coordinator. What do you think?

File: core/src/test/scala/unit/kafka/server/ListGroupsRequestTest.scala [hunk 24]
```diff
@@ -0,0 +1,236 @@
...
+        memberId = memberId2,
+        useNewProtocol = true
+      )
+
+      checkListedGroups(
+        groupId = "grp",
+        state = ConsumerGroupState.EMPTY.toString,
+        statesFilterExpectingEmptyListedGroups = List(ConsumerGroupState.RECONCILING.toString),
+        version = version
+      )
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ListGroupsRequestTest.scala]: I find this construct a bit weird because we mix two things into one. It may be better to separate. One way would be to list group and assert the returned list.

File: core/src/test/scala/unit/kafka/server/ListGroupsRequestTest.scala [hunk 25]
```diff
@@ -0,0 +1,236 @@
...
+      deleteGroups(
+        groupIds = List("grp"),
+        expectedErrors = List(Errors.NONE),
+        version = ApiKeys.DELETE_GROUPS.latestVersion(isUnstableApiEnabled)
+      )
+
+      assertEquals(
+        List.empty,
+        listGroups(
+          statesFilter = List.empty,
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ListGroupsRequestTest.scala:144]: Should we also have a test with multiple states? It seems that we only do empty or one.

File: core/src/test/scala/unit/kafka/server/OffsetDeleteRequestTest.scala [hunk 26]
```diff
@@ -0,0 +1,165 @@
...
+    createOffsetsTopic()
+
+    // Create the topic.
+    createTopic(
+      topic = "foo",
+      numPartitions = 3
+    )
+
+    for (version <- ApiKeys.OFFSET_DELETE.oldestVersion() to ApiKeys.OFFSET_DELETE.latestVersion(isUnstableApiEnabled)) {
+
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/OffsetDeleteRequestTest.scala]: nit: empty line.

File: core/src/test/scala/unit/kafka/server/OffsetDeleteRequestTest.scala [hunk 27]
```diff
@@ -0,0 +1,165 @@
...
+          memberId = memberId,
+          memberEpoch = memberEpoch,
+          subscribedTopicNames = List.empty
+        )
+      } else {
+        leaveGroup(
+          groupId = "grp",
+          memberId = memberId,
+          useNewProtocol = false
+        )
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/OffsetDeleteRequestTest.scala:123]: I wonder we could also test the "unsubscribe" case for the old protocol and the leave group case for the new one for completeness.

File: core/src/test/scala/unit/kafka/server/OffsetDeleteRequestTest.scala [hunk 28]
```diff
@@ -0,0 +1,165 @@
...
+      deleteOffset(
+        groupId = "grp",
+        topic = "foo",
+        partition = 5,
+        expectedPartitionError = Errors.UNKNOWN_TOPIC_OR_PARTITION,
+        version = version.toShort
+      )
+
+      // Delete offset with unknown group id.
+      deleteOffset(
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/OffsetDeleteRequestTest.scala:146]: Could we also add the case that led to https://github.com/apache/kafka/pull/14589? Could we also update the OffsetFetch integration tests while we are here?

**Review by dajac — COMMENTED** [review]
@dongnuo123 Thanks for the PR. I left some comments.

File: core/src/test/scala/unit/kafka/server/ListGroupsRequestTest.scala (hunk 23 above)
Comment (dongnuo123) [reply core/src/test/scala/unit/kafka/server/ListGroupsRequestTest.scala]: It may not be very related to the tests -- I realize that even it's both in STABLE states, we can't filter and get the groups in new protocol and old protocol at the same time. Is that the feature we want?

File: core/src/test/scala/unit/kafka/server/OffsetDeleteRequestTest.scala (hunk 28 above)
Comment (dongnuo123) [reply core/src/test/scala/unit/kafka/server/OffsetDeleteRequestTest.scala:146]: Can we add those cases after merging #14589, currently tests for those cases won't pass

File: core/src/test/scala/unit/kafka/server/OffsetDeleteRequestTest.scala (hunk 27 above)
Comment (dongnuo123) [reply core/src/test/scala/unit/kafka/server/OffsetDeleteRequestTest.scala:123]: I don't know how to unsubscribe a topic in the old protocol except making all members subscribing to that topic leave..

File: core/src/test/scala/unit/kafka/server/OffsetDeleteRequestTest.scala (hunk 28 above)
Comment (dongnuo123) [reply core/src/test/scala/unit/kafka/server/OffsetDeleteRequestTest.scala:146]: new tests added.

File: core/src/test/scala/unit/kafka/server/OffsetDeleteRequestTest.scala [hunk 29]
```diff
@@ -0,0 +1,182 @@
...
+        topic = "foo-unknown",
+        partition = 2,
+        expectedResponseError = Errors.INVALID_GROUP_ID,
+        version = version.toShort
+      )
+
+      // Delete offset with both invalid id and invalid partition should return with a response with top-level error.
+      deleteOffset(
+        groupId = "",
+        topic = "foo-unknown",
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/OffsetDeleteRequestTest.scala]: Don't we need to use `foo` here?

File: core/src/test/scala/unit/kafka/server/OffsetDeleteRequestTest.scala [hunk 30]
```diff
@@ -0,0 +1,182 @@
...
+      deleteOffset(
+        groupId = "",
+        topic = "foo",
+        partition = 2,
+        expectedResponseError = Errors.INVALID_GROUP_ID,
+        version = version.toShort
+      )
+
+      // Delete offset with both invalid id and invalid topic should return with a response with top-level error.
+      deleteOffset(
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/OffsetDeleteRequestTest.scala:164]: Could we also add similar tests for the offset fetch as we had a similar issue there?

**Review by dajac — APPROVED** [review]
LGTM, thanks.
